### PR TITLE
EVP: fix keygen for EVP_PKEY_RSA_PSS

### PR DESCRIPTION
**Pull request by danbev**
This commit attempts to fix the an issue when generating a key of type
`EVP_PKEY_RSA_PSS`. Currently, `EVP_PKEY_CTX_set_rsa_keygen_bits` will
return -1 if the key id is not of type `EVP_PKEY_RSA`. This commit adds
`EVP_PKEY_RSA_PSS` to also be accepted.

After applying the above change, calling
`EVP_PKEY_CTX_set_rsa_pss_keygen_md(ctx, md)` will fail. The link below
contains details about this. As mentioned in the linked doc I'm not
sure at all about the change, if it is correct or done in the right
place but I hope this can be a starting point for a discussion.

Fixes: https://github.com/openssl/openssl/issues/12384
Refs: https://github.com/danbev/learning-libcrypto/blob/master/notes/issues.md#evp_pkey_ctx_set_rsa_pss_keygen_md-issue

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
https://github.com/openssl/openssl/blob/3094351625f0b222f92c22ce4943461df8c7e301/include/openssl/rsa.h#L169-L172

This is very wrong. This has not been converted to be "provider" aware. Its the old 1.1.1 way of doing things. It needs to be converted into a function along the lines of some of the other "md" setting functions, e.g.

https://github.com/openssl/openssl/blob/3094351625f0b222f92c22ce4943461df8c7e301/crypto/rsa/rsa_lib.c#L900-L958

**Comment by danbev** [comment]
@mattcaswell Thanks for the feedback, I'll take a stab at converting it into a function on Monday.

**Comment by danbev** [comment]
@mattcaswell Would you be able to take a look at the changes in https://github.com/openssl/openssl/pull/13099/commits/54748a045ca8d9850b3934c1a1eb7741fe76d74b which my attempt at converting the macro into a function.

File: crypto/evp/pmeth_lib.c
Comment (levitte) [inline crypto/evp/pmeth_lib.c:1344]: Why this direct call instead of `EVP_PKEY_CTX_set_rsa_pss_keygen_md(ctx, p2)`?

Comment (mattcaswell) [reply crypto/evp/pmeth_lib.c:1344]: This is incorrect. This function is called `legacy_ctrl_to_param`. Its purpose is to convert instances where the user has made legacy ctrl calls, and they need to be converted into new style params calls. This is just calling the legacy ctrl instead.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Looks good apart from the one issue noted (also noted by @levitte)

File: crypto/evp/pmeth_lib.c
Comment (danbev) [reply crypto/evp/pmeth_lib.c:1344]: Sorry about that, I'll take a look now. Thanks

File: crypto/rsa/rsa_lib.c
Comment (levitte) [inline crypto/rsa/rsa_lib.c:915]: Ok, it's looking better.  However, we do need legacy support *somewhere* (because engines), and this would be the spot, as far as I can tell.  Here's the code that I would insert, right here:

``` C
    /* TODO(3.0): Remove this eventually when no more legacy */
    if (ctx->op.keymgmt.genctx == NULL)
        return EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_RSA_PSS, EVP_PKEY_OP_KEYGEN,
                                 EVP_PKEY_CTRL_MD, 0, (void *)md);
```

**Review by levitte — CHANGES_REQUESTED** [review]
Last bit missing, the connection back to legacy.

File: crypto/rsa/rsa_lib.c
Comment (levitte) [reply crypto/rsa/rsa_lib.c:915]: Er, I would insert that *before* the `name` assignment.

Comment (danbev) [reply crypto/rsa/rsa_lib.c:920]: Thanks, i was not sure how to handle the legacy case. I've update now.

File: crypto/rsa/rsa_lib.c
Comment (levitte) [inline crypto/rsa/rsa_lib.c:912]: We usually break lines before the logical operator, so something like this would be preferable:

``` C
    if (ctx->pmeth != NULL && ctx->pmeth->pkey_id != EVP_PKEY_RSA
        && ctx->pmeth->pkey_id != EVP_PKEY_RSA_PSS)
```

or

``` C
    if (ctx->pmeth != NULL
        && ctx->pmeth->pkey_id != EVP_PKEY_RSA
        && ctx->pmeth->pkey_id != EVP_PKEY_RSA_PSS)
```

File: crypto/rsa/rsa_lib.c
Comment (levitte) [inline crypto/rsa/rsa_lib.c:939]: Same here

**Review by levitte — CHANGES_REQUESTED** [review]
Nits

File: crypto/rsa/rsa_lib.c
Comment (danbev) [reply crypto/rsa/rsa_lib.c]: Thanks, I've updated this now.

File: crypto/rsa/rsa_lib.c
Comment (danbev) [reply crypto/rsa/rsa_lib.c]: Updated now.

**Comment by levitte** [comment]
Current CI failures are not relevant for this PR.  They are fixed in #13640

**Comment by t8m** [comment]
In case of commits that just fix issues found by review that do not add anything substantially new, I'd recommend using --fixup instead of --squash. (Just FYI for future commits.)

**Comment by levitte** [comment]
> In case of commits that just fix issues found by review that do not add anything substantially new, I'd recommend using --fixup instead of --squash. (Just FYI for future commits.)

"... substantially new **in commit messages** ...", I might add.

**Comment by t8m** [comment]
I meant that commit message `Fix line breaks in if statements.` is not something that should be there in the squashed commit because the commit to be squashed is actually fixing stuff that was added by the primary commit. So in the end the message is superfluous and misleading.

In general squash (instead of fixup) commits are not that useful because if you are adding some new unrelated or just remotely related fix you can just use a separate non-squashed commit and if you're redoing the fix in substantial way that requires change of the original commit message it is usually better to just squash and force-push on your branch already to amend the original commit.

**Comment by levitte** [comment]
> In general squash (instead of fixup) commits are not that useful because if you are adding some new unrelated or just remotely related fix you can just use a separate non-squashed commit and if you're redoing the fix in substantial way that requires change of the original commit message it is usually better to just squash and force-push on your branch already to amend the original commit.

There are different views on this.  Some that do regular reviews, and that includes me, like to be able to see what progress was made since the last review, and fixup / squash commits are often very easy.  Viewing the force-push sometimes works, and sometimes not, depending on if a rebase was made at the same time or not.

**Comment by danbev** [comment]
Is there a way for me to trigger a re-run of the tests? I've rebased this PR but only the `appveyor` job seems to have been run for it.

**Comment by danbev** [comment]
@levitte @t8m Thanks for all your help on this! Please let me know if there is anything I can do to get this merged.

**Review by t8m — CHANGES_REQUESTED** [review]
Please drop the rename of the OSSL_PKEY_PARAM_RSA_DIGEST.

File: crypto/rsa/rsa_lib.c
Comment (danbev) [reply crypto/rsa/rsa_lib.c]: Thanks, I'll remove that.

File: crypto/evp/pmeth_lib.c
Comment (danbev) [reply crypto/evp/pmeth_lib.c]: For some reason I though there should be a separate one for RSA-PSS. I'll remove these shortly. Thanks

**Review by t8m — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
ping for the second review

**Comment by t8m** [comment]
As this stalled in the review process and there was a conflict in libcrypto.num I've rebased this to fix the conflict and squashed the commits. No other changes were done. @openssl/committers please provide the second review.

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master, thank you for the contribution and patience with the reviews.

**Pull request closed without merging**
